src/main/java/com/google/sps/utility/LocationsUtility.java

```diff
@@ -63,4 +64,39 @@ private static String getLocation(Pattern pattern, String prefix, String taskNot
     }
     return "";
   }
+
+  /**
+   * Generates permutations of the lists and appends them to the result parameter.
+   *
+   * @param lists A list of lists to generate permutations for.
+   * @param result A pointer to a list to contain the generated permutations.
+   * @param depth An int to represent the index of the list that is currently being traversed in the
+   *     recursive call.
+   * @param current A list containing the permutation built so far at the recursive call.
+   */
+  private static void generateCombinationsHelper(
+      List<List<String>> lists, List<List<String>> result, int depth, List<String> current) {
+    if (depth == lists.size()) {
+      result.add(new ArrayList<>(current));
+      return;
+    }
+    for (int i = 0; i < lists.get(depth).size(); i++) {
+      List<String> next = new ArrayList<>();
+      next.addAll(current);
+      next.add(lists.get(depth).get(i));
+      generateCombinationsHelper(lists, result, depth + 1, next);
+    }
+  }
+
+  /**
+   * Generates all permutations of the lists recursively. Scope of method is public for testing
+   * purposes.
+   *
+   * @param lists A list of lists to generate permutations for.
+   */
+  public static List<List<String>> generateCombinations(List<List<String>> lists) {
+    List<List<String>> combinations = new ArrayList<>();
+    generateCombinationsHelper(lists, combinations, 0, new ArrayList<>());
+    return combinations;
+  }
 }
```

Comment:
I'm confused by what you mean by combination and permutation. When you say "Generates all permutations of the lists recursively", I would expect the method to return a list with lists.length! elements, where each element would be a possible permutation of the elements of `lists`. Could you please clarify?

Also, why is this being done?

Reply:
Do you mean all possible combinations containing a single element from each list?

Reply:
Oh I meant combinations for everything, thanks for catching that.

Reply:
Not permutations because order does not matter.